File: Tests/iOS Tests/PNPublishCompressedTests.m
//
//  PNPublishCompressedTests.m
//  PubNub Tests
//
//  Created by Vadim Osovets on 6/22/15.
//
//

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import <PubNub/PubNub.h>

#import "PNBasicClientTestCase.h"

#import "NSString+PNTest.h"

@interface PNPublishCompressedTests : PNBasicClientTestCase

@end

@implementation PNPublishCompressedTests

- (BOOL)recording {
    return NO;
    
}

- (void)testSimplePublishCompressed {
    [self performVerifiedPublish:@"test"
                       onChannel:[NSUUID UUID].UUIDString
                      compressed:YES withAssertions:^(PNStatus<PNPublishStatus> *status) {
                          XCTAssertNotNil(status);
                          XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                          XCTAssertEqual(status.operation, PNPublishOperation);
                          XCTAssertEqual(status.statusCode, 200);
                          XCTAssertFalse(status.isError);
                          NSLog(@"status.data.information: %@", status.data.information);
                          NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                          XCTAssertEqualObjects(status.data.information, @"Sent");
                      }];
}

- (void)testSimplePublishNotCompressed {
    [self performVerifiedPublish:@"test"
                       onChannel:[NSUUID UUID].UUIDString
                      compressed:NO
                  withAssertions:^(PNStatus<PNPublishStatus> *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                  }];
}

- (void)testPublishNilMessageCompressed {
    [self performVerifiedPublish:nil
                       onChannel:[NSUUID UUID].UUIDString
                      compressed:YES
                  withAssertions:^(PNStatus<PNPublishStatus> *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNBadRequestCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 400);
                      XCTAssertTrue(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertNil(status.data.information);
                      XCTAssertNil(status.data.timetoken);
                  }];
}

- (void)testPublishNilMessageNotCompressed {
    [self performVerifiedPublish:nil
                       onChannel:[NSUUID UUID].UUIDString
                      compressed:NO
                  withAssertions:^(PNStatus<PNPublishStatus> *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNBadRequestCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 400);
                      XCTAssertTrue(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertNil(status.data.information);
                      XCTAssertNil(status.data.timetoken);
                  }];
}

- (void)testPublishDictionaryCompressed {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:[NSUUID UUID].UUIDString
                      compressed:YES
                  withAssertions:^(PNStatus<PNPublishStatus> *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                  }];
}

- (void)testPublishDictionaryNotCompressed {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:[NSUUID UUID].UUIDString
                      compressed:NO
                  withAssertions:^(PNStatus<PNPublishStatus> *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                  }];
}

- (void)testPublishToNilChannelCompressed {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:nil
                      compressed:YES
                  withAssertions:^(PNStatus<PNPublishStatus> *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNBadRequestCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 400);
                      XCTAssertTrue(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertNil(status.data.information);
                      XCTAssertNil(status.data.timetoken);
                  }];
}

- (void)testPublishToNilChanneNotCompressed {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:nil
                      compressed:NO
                  withAssertions:^(PNStatus<PNPublishStatus> *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNBadRequestCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 400);
                      XCTAssertTrue(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertNil(status.data.information);
                      XCTAssertNil(status.data.timetoken);
                  }];
}

#pragma mark - Main flow

- (void)performVerifiedPublish:(id)message
                     onChannel:(NSString *)channel
                    compressed:(BOOL)compressed
                withAssertions:(PNPublishCompletionBlock)verificationBlock {
    XCTestExpectation *networkExpectation = [self expectationWithDescription:@"network"];
    [self.client publish:message toChannel:channel compressed:compressed
          withCompletion:^(PNStatus<PNPublishStatus> *status) {
              verificationBlock(status);
              [networkExpectation fulfill];
          }];
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

@end

File: Tests/iOS Tests-Swift/PNPublishCompressedTests.swift
//
//  PNPublishCompressedTests.swift
//  PubNub Tests
//
//  Created by Vadim Osovets on 09/14/15.
//
//

import UIKit
import XCTest

class PNPublishCompressedTests: PNBasicClientTestCase {
    
    let publishChannelString = "F16CB07C-9F3F-41AA-8A0A-313960F21AAB"
    
    override func isRecording() -> Bool {
        return false
    }
    
    func testSimplePublishCompressed() {
        self.performVerifiedPublish("test", onChannel: self.publishChannelString, compressed: true) { (status: PNPublishStatus!) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information as String, "Sent")
            XCTAssertEqual(status.data.timetoken, NSNumber(longLong:14355315263325276))
        }
    }
    
    func testSimplePublishNotCompressed() {
        self.performVerifiedPublish("test", onChannel: self.publishChannelString, compressed: false) { (status: PNPublishStatus!) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information as String, "Sent")
            XCTAssertEqual(status.data.timetoken, NSNumber(longLong:14355315264254610))
        }
    }
    
    
    func testPublishNilMessageCompressed() {
        
        let localTestExpectation = self.expectationWithDescription("network")
        
        self.client.publish(nil, toChannel:self.publishChannelString, compressed: true) { (status) -> Void in

            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNBadRequestCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 400)
            XCTAssertTrue(status.error)
            XCTAssertNil(status.data)
            
            localTestExpectation.fulfill()
        }
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testPublishNilMessageNotCompressed() {
        
        let networkExpectation = self.expectationWithDescription("network")
        
        self.client.publish(nil, toChannel:self.publishChannelString, compressed: false) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNBadRequestCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 400)
            XCTAssertTrue(status.error)
            XCTAssertNil(status.data)
            
            networkExpectation.fulfill()
        }
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testPublishDictionaryCompressed() {
        self.performVerifiedPublish(["test" : "test"], onChannel: self.publishChannelString, compressed: true) { (status: PNPublishStatus!) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information as String, "Sent")
            XCTAssertEqual(status.data.timetoken, NSNumber(longLong:14355315261129449))
        }
    }
    
    func testPublishDictionaryNotCompressed() {
        self.performVerifiedPublish(["test" : "test"], onChannel: self.publishChannelString, compressed: false) { (status: PNPublishStatus!) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information as String, "Sent")
            XCTAssertEqual(status.data.timetoken, NSNumber(longLong:14355315262089406))
        }
    }
    
    func testPublishToNilChannelCompressed() {
        
        let networkExpectation = self.expectationWithDescription("network")
        
        self.client.publish(["test" : "test"], toChannel:nil, compressed: true) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNBadRequestCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 400)
            XCTAssertTrue(status.error)
            XCTAssertNil(status.data)
            
            networkExpectation.fulfill()
        }
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
    
    func testPublishToNilChanneNotCompressed() {
        
        let networkExpectation = self.expectationWithDescription("network")
        
        self.client.publish(["test" : "test"], toChannel:nil, compressed: false) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNBadRequestCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 400)
            XCTAssertTrue(status.error)
            XCTAssertNil(status.data)
            
            networkExpectation.fulfill()
        }
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
}

extension PNPublishCompressedTests {
    func performVerifiedPublish(message: AnyObject, onChannel:String,
        compressed: Bool, withAssertions: PNPublishCompletionBlock) {
            
            let networkExpectation = self.expectationWithDescription("network")
            
            self.client.publish(message, toChannel: onChannel, compressed: compressed) { (status) -> Void in
                withAssertions(status)
                networkExpectation.fulfill()
            }
            waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
                XCTAssertNil(error, "Encountered error with publish call")
            })
    }
}
